doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,205 @@
+
+Single method typeclasses are cheap because they have simpler representations in GHC Core - a newtype that can be compiled away.
+On the other hand, multiple method typeclasses exist in GHC Core as actual dictionaries and need to be compiled into product types in PIR.
+
+This makes monomorphization the only choice in many cases. Consider `foldr` and `foldMap`.
+They can be implemented in terms of each other, so technically, only one of them needs to be in the `Foldable` class.
+But certain types have more efficient direct implementations of `foldr`; the same for `foldMap`.
+No matter which of them we move out of `Foldable`, it won't work optimally for certain types.
+Keeping both in `Foldable` will cause `Foldable` to be a multiple method typeclass, which is expensive in and of itself, as explained above.
+We may as well do away with `PlutusTx.Foldable`, and instead simply provide monomorphic versions of `Foldable`-related functions for each common type.
+
+### Turning off GHC's pre-inliner
+
+As said before, one reason why the semantics of Plutus Tx is difficult to predict is because we run GHC's pre-inliner, which may perform unconditionally inlining.
+Since GHC 9.0, GHC provides [an option](https://hackage.haskell.org/package/ghc-9.6.1/docs/GHC-Core-Opt-Simplify-Env.html#v:sm_pre_inline) to turn it off when running the simplifier.
+By turning it off, GHC (hopefully) won't inline anything at all, which means it won't change the strictness of Plutus Tx programs.
+
+At this time we can't simply turn the pre-inliner off, because doing so breaks the Plutus Tx compiler.
+For details, see [this note](https://github.com/input-output-hk/plutus/blob/dc57fe1b8497835ef2a7794c017c5a50c0b7e647/plutus-tx-plugin/src/PlutusTx/Plugin.hs#L123).
+The problem described in the note should, however, be resolved in a different way, as explained in the note.
+Once done, we can go ahead and turn it off.
+
+### Performing GHC-like Inlining for Functions with the INLINE Pragma
+
+In Plutus Tx, like most other strict languages, only a few built-in constructs are non-strict (such as if-then-else).
+It is cumbersome, if not impossible, to create user-defined non-strict functions.
+The `nonstrict` function described before does not help because it cannot be used in defining a function.
+It would be useful to allow user-defined non-strict functions, so that we can define functions like `&&` and `||` that short-circuit properly.
+Some languages like Scala have special syntaxes for defining non-strict functions; we can't change Haskell's syntax, so we need to find a different solution.
+
+The idea is to make the Plutus Tx compiler perform GHC-like inlining for functions with the INLINE pragma.
+
+"GHC-like inlining", or "inlining in the GHC way", for lack of a better term, refers to inlining in a way similar to how GHC does it, i.e., replacing a variable with its RHS, then performing beta substitution.
```

Comment:
I'm confused by this. I don't think GHC's inliner actually does the substitution? I think the simplifier will do it, but I find it confusing to associate this with GHC's inlining somehow?

Reply:
I agree. I'm not sure how GHC draws the line exactly. To avoid confusion maybe we can just talk about the two ways of inlining explicitly:

(1) inlining AND performing the substitution
(2) only the inlining

Reply:
This is mainly a terminology issue. but look at the example below: by "GHC-like inlining", I mean getting the second result, which is what we'll get if we run GHC's inliner. Our own inliner will only give us the first result. Please suggest better terminology.

Reply:
I think what you're seeing there is due to the fact that GHC's inliner is just a part of the simplifier. So it's getting both inlined and simplified. I think it's confusing to call this "GHC-style inlining" since it isn't IMO. It's inlining-followed-by-semantically-incorrect-for-PIR-beta-reduction :joy: 

What I think this most looks like is macro-expansion. Indeed, that's what I've suggested in the past as a way to get short-circuiting behaviour: use macros (i.e. TH). You can [very easily](https://github.com/input-output-hk/plutus/pull/4191#issuecomment-962004503) write a TH macro that expands inline the way you want it. And it's much less misleading, since it also clearly only works when you have the argument expressions to hand. Whereas your `(&&)` can be passed e.g. to a higher-order-function where it will _not_ behave lazily with no indication of that. 

As you can tell I'm pretty unconvinced by this proposal. I think it abuses the meaning of `INLINE`, and it's fragile and unreliable.